sdk/storage/Azure.Storage.Files.Shares/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,392 @@
+            }
+
+            // Set _rangeSize
+            if (transferOptions.MaximumTransferSize.HasValue
+                && transferOptions.MaximumTransferSize.Value > 0)
+            {
+                _rangeSize = Math.Min(transferOptions.MaximumTransferSize.Value, Constants.Blob.Block.MaxDownloadBytes);
+            }
+            else
+            {
+                _rangeSize = Constants.DefaultBufferSize;
+            }
+
+            // Set _initialRangeSize
+            if (transferOptions.InitialTransferSize.HasValue
+                && transferOptions.InitialTransferSize.Value > 0)
+            {
+                _initialRangeSize = transferOptions.InitialTransferSize.Value;
+            }
+            else
+            {
+                _initialRangeSize = Constants.Blob.Block.DefaultInitalDownloadRangeSize;
+            }
+        }
+
+        public async Task<Response> DownloadToAsync(
+            Stream destination,
+            ShareFileRequestConditions conditions,
+            CancellationToken cancellationToken)
+        {
+            // Wrap the download range calls in a Download span for distributed
+            // tracing
+            DiagnosticScope scope = _client.ClientConfiguration.ClientDiagnostics.CreateScope($"{nameof(ShareFileClient)}.{nameof(ShareFileClient.DownloadTo)}");
+            try
+            {
+                scope.Start();
+
+                // Just start downloading using an initial range.  If it's a
+                // small file, we'll get the whole thing in one shot.  If it's
+                // a large file, we'll get its full size in Content-Range and
+                // can keep downloading it in segments.
+                var initialRange = new HttpRange(0, _initialRangeSize);
+                Task<Response<ShareFileDownloadInfo>> initialResponseTask =
+                    _client.DownloadAsync(
+                        initialRange,
+                        rangeGetContentHash: false,
+                        conditions,
+                        cancellationToken);
+
+                Response<ShareFileDownloadInfo> initialResponse = null;
+                try
+                {
+                    initialResponse = await initialResponseTask.ConfigureAwait(false);
+                }
+                catch (RequestFailedException ex) when (ex.ErrorCode == ShareErrorCode.InvalidRange)
+                {
+                    initialResponse = await _client.DownloadAsync(
+                        range: default,
+                        false,
+                        conditions,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                }
+
+                // If the initial request returned no content (i.e., a 304),
+                // we'll pass that back to the user immediately
+                if (initialResponse.IsUnavailable())
+                {
+                    return initialResponse.GetRawResponse();
+                }
```

Comment:
We got a github issue about this a few weeks ago, I wonder if we should remove this here too.

Reply:
Could you link the issue? I can't seem to find it.

Reply:
https://github.com/Azure/azure-sdk-for-net/pull/22647
It was about not having that block... 
Please make sure to add tests that capture 304 response.